fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java

```diff
@@ -161,16 +167,48 @@ public GenericResultsetData retrieveGenericResultset(final String name, final St
 
         final long startTime = System.currentTimeMillis();
         LOG.info("STARTING REPORT: {}   Type: {}", name, type);
-
-        final String sql = getSQLtoRun(name, type, queryParams, isSelfServiceUserReport);
-
-        final GenericResultsetData result = this.genericDataService.fillGenericResultSet(sql);
+        final StringBuilder sqlStringBuilder = new StringBuilder(200);
+        sqlStringBuilder.append(getSQLtoRun(name, type, queryParams, isSelfServiceUserReport));
+        final GenericResultsetData result;
+        boolean isPaginationAllowed = Boolean.parseBoolean(queryParams.get(ReportingConstants.IS_PAGINATION_ALLOWED));
+
+        if (isPaginationAllowed) {
+            result = retrieveGenericResultsetWithPagination(sqlStringBuilder, queryParams);
```

Comment:
Let's not pass a mutable StringBuilder as a parameter.
The code should be looking like this:
```
String sql = ... // base SQL
...
if (pagination) {
   sql = retrievePaginatedSql(sql)
}
...
return fillGenericResultset(sql)
```

Or similar to this.